testsuite/testsuite_impl/my_test/my_test_.go

```diff
@@ -0,0 +1,202 @@
+	}
+	logrus.Infof("Logs: %+v", string(*logOutput))
+
+	exitCode, logOutput, err = serviceCtx.ExecCommand([]string{"/bin/sh", "-c",
+		fmt.Sprintf("geth attach /tmp/geth.ipc --exec 'eth.getBalance(eth.accounts[1])'"),
+	})
+	if err != nil || exitCode != 0 {
+		return stacktrace.NewError("Executing command returned either an error or failing exit code with logs: %+v", string(*logOutput))
+	}
+	logrus.Infof("Logs: %+v", string(*logOutput))
+	return nil
+}
+
+// ====================================================================================================
+//                                       Private helper functions
+// ====================================================================================================
+func getMyServiceConfigurations() (*services.ContainerCreationConfig, func(ipAddr string, generatedFileFilepaths map[string]string, staticFileFilepaths map[services.StaticFileID]string) (*services.ContainerRunConfig, error)) {
+	containerCreationConfig := getContainerCreationConfig()
+
+	runConfigFunc := getRunConfigFunc()
+	return containerCreationConfig, runConfigFunc
+}
+
+func getContainerCreationConfig() *services.ContainerCreationConfig {
+	containerCreationConfig := services.NewContainerCreationConfigBuilder(
+		ethereumGoClientImageName,
+	).WithUsedPorts(
+		map[string]bool{fmt.Sprintf("%v/tcp", ethereumGoClientPort): true},
+	).Build()
+	return containerCreationConfig
+}
+
+func getRunConfigFunc() func(ipAddr string, generatedFileFilepaths map[string]string, staticFileFilepaths map[services.StaticFileID]string) (*services.ContainerRunConfig, error) {
+	runConfigFunc := func(ipAddr string, generatedFileFilepaths map[string]string, staticFileFilepaths map[services.StaticFileID]string) (*services.ContainerRunConfig, error) {
+		entrypointCommand := fmt.Sprintf("geth --dev -http --http.api admin,eth,net,rpc --http.addr %v --http.corsdomain '*' --nat extip:%v",
+			ipAddr,
+			ipAddr)
+		entrypointArgs := []string{
+			"/bin/sh",
+			"-c",
+			entrypointCommand,
+		}
+		result := services.NewContainerRunConfigBuilder().WithEntrypointOverride(entrypointArgs).Build()
+		return result, nil
+	}
+	return runConfigFunc
+}
+
+func getEnodeAddress(ipAddress string) (string, error) {
+	nodeInfoResponse := new(NodeInfoResponse)
+	err := sendRpcCall(ipAddress, adminInfoRpcCall, nodeInfoResponse)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "Failed to send admin node info RPC request to geth node.")
+	}
+	return nodeInfoResponse.Result.Enode, nil
+}
+
+func sendRpcCall(ipAddress string, rpcJsonString string, targetStruct interface{}) error {
+	url := fmt.Sprintf("http://%v:%v", ipAddress, rpcPort)
+	var jsonByteArray = []byte(rpcJsonString)
+
+	resp, err := http.Post(url, "application/json", bytes.NewBuffer(jsonByteArray))
+	if err != nil {
+		return stacktrace.Propagate(err, "Failed to send RPC request to geth node")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
```

Comment:
We can do better here with "eject early, eject often" - you can invert this and it'll be a lot more concise:
```
if resp.StatusCode != http.StatusOk {
    return stacktrace.NewError("Received non-200 status code from admin RPC API: %v", resp.StatusCode)
}
```

Reply:
oh yes! I have to get used to that pattern, thanks.